### PR TITLE
feat(crypto): CRP-2563 Support key derivation in ic_crypto_ecdsa_secp256k1

### PR DESCRIPTION
**Pull request by randombit**
Both public and private key derivation are supported. This implementation matches the extended BIP32 used on the IC for threshold ECDSA signatures. This allows users either within or outside of the monorepo to perform this operation without relying on our internal implementation.

### CODE REVIEW

**Review by basvandijk — APPROVED** [review]

File: rs/crypto/ecdsa_secp256k1/tests/tests.rs [hunk 1]
```diff
@@ -259,15 +284,67 @@ i389XZmdlKFbsLkUI9dDQgMP98YnUA==
     );
 }
 
+#[test]
+fn should_match_slip10_derivation_test_data() {
```
Comment (altkdf) [inline rs/crypto/ecdsa_secp256k1/tests/tests.rs:288]: please add a link to the source of the test data

**Review by altkdf — COMMENTED** [review]
Thanks for adding the non-internal implementation @randombit! I have a few minor questions/suggestions.

* It could be useful to add some very short information to the PR description on why this is useful. Currently the description only tells what is being added.
* CRP-2564 in the title should be CRP-2563.
* It seems to me that having a differential fuzzer for duplicated code could be useful, WDYT?

File: rs/crypto/ecdsa_secp256k1/tests/tests.rs [hunk 2]
```diff
@@ -1,5 +1,20 @@
...
+use rand_chacha::ChaCha20Rng;
+
+fn test_rng_with_seed(seed: [u8; 32]) -> ChaCha20Rng {
+    use rand::SeedableRng;
+    ChaCha20Rng::from_seed(seed)
+}
+
+fn test_rng() -> ChaCha20Rng {
+    let seed = rand::thread_rng().gen::<[u8; 32]>();
+    // If a test ever fails, reproduce it using
```
[the remark this answers is not in the document]
Comment (randombit) [reply rs/crypto/ecdsa_secp256k1/tests/tests.rs:13]: Yes but it's also 5 lines of trivial code that leaves this crate depending only on published `crates.io` crates

File: rs/crypto/ecdsa_secp256k1/src/lib.rs [hunk 3]
```diff
@@ -513,6 +671,40 @@ impl PublicKey {
...
+        let public_key: AffinePoint = *self.key.as_affine();
+        let (pt, _offset, chain_code) = derivation_path.derive_offset(public_key, chain_code);
+
+        let derived_key = Self {
+            key: k256::ecdsa::VerifyingKey::from(
+                k256::PublicKey::from_affine(pt).expect("Derived point is valid"),
+            ),
+        };
+
+        (derived_key, chain_code)
```
[the remark this answers is not in the document]
Comment (randombit) [reply rs/crypto/ecdsa_secp256k1/src/lib.rs:717]: Done

File: rs/crypto/ecdsa_secp256k1/src/lib.rs [hunk 4]
```diff
@@ -329,6 +448,45 @@ impl PrivateKey {
         let key = self.key.verifying_key();
         PublicKey { key: *key }
     }
+
+    /// Derive a private key from this private key using a derivation path
+    ///
+    /// This is the same derivation system used by the Internet Computer when
```
[the remark this answers is not in the document]
Comment (randombit) [reply rs/crypto/ecdsa_secp256k1/src/lib.rs:454]: Done

**Review by altkdf — APPROVED** [review]

**Pull request merged**